nuxeo-features/nuxeo-search/nuxeo-search-core/src/main/java/org/nuxeo/ecm/platform/search/core/SavedSearchServiceImpl.java

```diff
@@ -0,0 +1,163 @@
+        savedSearch.setQuery(query);
+        savedSearch.setQueryLanguage(queryLanguage);
+        savedSearch.setPageProviderName(pageProviderName);
+        savedSearch.setPageSize(pageSize);
+        savedSearch.setCurrentPageIndex(currentPageIndex);
+        savedSearch.setMaxResults(maxResults);
+        savedSearch.setSortBy(sortBy);
+        savedSearch.setSortOrder(sortOrder);
+        savedSearch.setContentViewData(contentViewData);
+
+        savedSearchDoc = session.createDocument(savedSearchDoc);
+        savedSearch = savedSearchDoc.getAdapter(SavedSearch.class);
+
+        ACP acp = savedSearchDoc.getACP();
+        ACL acl = acp.getOrCreateACL(ACL.LOCAL_ACL);
+        Principal principal = session.getPrincipal();
+        if (principal != null) {
+            acl.add(new ACE(principal.getName(), SecurityConstants.EVERYTHING, true));
+        }
+
+        acp.addACL(acl);
+        savedSearchDoc.setACP(acp, true);
+
+        return savedSearch;
+    }
+
+    @Override
+    public SavedSearch getSavedSearch(CoreSession session, String id) {
+        DocumentRef docRef = new IdRef(id);
+        DocumentModel savedSearchDoc = session.getDocument(docRef);
+        if (savedSearchDoc != null) {
+            SavedSearch search = savedSearchDoc.getAdapter(SavedSearch.class);
+            if (search != null) {
+                return search;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public SavedSearch saveSavedSearch(CoreSession session, SavedSearch search) throws InvalidSearchParameterException,
+            IOException {
+        if (Strings.isNullOrEmpty(search.getTitle())) {
+            throw new InvalidSearchParameterException("title cannot be empty");
+        }
+
+        if ((!Strings.isNullOrEmpty(search.getQuery()) || !Strings.isNullOrEmpty(search.getQueryLanguage()))
+                && !Strings.isNullOrEmpty(search.getPageProviderName())) {
+            throw new InvalidSearchParameterException("query and page provider parameters are mutually exclusive"
+                    + " (query, queryLanguage, pageProviderName)");
+        }
+
+        if (Strings.isNullOrEmpty(search.getQuery()) && Strings.isNullOrEmpty(search.getQueryLanguage())
+                && Strings.isNullOrEmpty(search.getPageProviderName())) {
+            throw new InvalidSearchParameterException("query or page provider parameters are missing"
+                    + " (query, queryLanguage, pageProviderName)");
+        }
+
+        if (!Strings.isNullOrEmpty(search.getQuery()) && Strings.isNullOrEmpty(search.getQueryLanguage())) {
+            throw new InvalidSearchParameterException("queryLanguage parameter is missing");
+        }
+
+        if (Strings.isNullOrEmpty(search.getQuery()) && !Strings.isNullOrEmpty(search.getQueryLanguage())) {
+            throw new InvalidSearchParameterException("query parameter is missing");
+        }
+
+        session.saveDocument(search.getDocument());
```

Comment:
Shouldn't the result of `saveDocument` be put back into the `search.setDocument()` ?